[2단계 - 자동차 경주] 몰리(김지민) 미션 제출합니다.

jminkkk opened the pull request:
안녕하세요 범블비 :)
2단계는 제가 1단계를 진행하면서 헷갈렸던 부분들의 이유를 찾고자 했어요!

제가 고민했던 부분들은 
1. CarController에서 전역변수로 List<String> names와 int moveCount를 가지고 있는 것이 찝찝한데 어떻게 해야 좋을까?
2. 각 자동차들의 이동 내역을 출력할 때, CarGroup 객체를 넘겨 toString을 출력하도록 구현이 되어있는데요. 이렇게 하는 경우 내부적으로는 실행 결과라는 멘트가 출력되고 나서 경주가 진행되는데, 이 흐름이 자연스러운 것일까?
3. InputView에서는 입력에 대한 책임만 가지고 있어야 하므로 입력받은 String 값을 그대로 컨트롤러에 넘기는 것이 맞지 않을까? 그 후에 NameParser를 통해 가공을 하는 것이 맞지 않을까?
4. RandomNumberGenerator를 CarGroup에서 가지고 있는 것이 맞을까?

제가 내린 결론은 
1. 전역변수를 지양하자.
내가 예상하지 못한 부분에서 값이 변경될 가능성이 있다.

2. MoveHistory라는 Record를 통해 경주가 진행될 경우 각 자동차들의 이동 내역을 저장하자. 
게임이 종료된 후에 모든 이동 내역을 결과와 함께 출력하도록 하는 것이 자연스러운 흐름같다.

3. List<String> 으로 파싱까지의 책임을 View에서 맡도록 하자. 
현재는 콘솔 입력이지만 요구사항이 변경될 경우, Controller에서 다른 형태의 입력값을 List<String>를 처리하기 위해 로직을 변경해야 한다. 따라서 InputView에서 List<String>로 반환하여 이러한 변경에 대처할 수 있도록 하자.

4. Controller에서 RandomNumberGenerator를 가지고 있자.
내가 제어할 수 없는 RandomNumberGenerator를 CarGroup에서 가지게 되면 단위테스트를 하기에 힘들 수 있다. 최대한 모델에서 제어할 수 없는 부분을 몰아내자.

추가로 궁금한 부분은 아래와 같습니다!

제가 이번에 이해한 일급 컬렉션은 불변하다는 장점이 있는 것인데요. 그렇다면 제가 기존에 구현한 addCar() 메서드가 있는 CarGroup은 일급 컬렉션으로 보기에 무리가 있는 것일까요?

hw0603 committed:
feat(Car): Car 클래스 생성

hw0603 committed:
feat(InputView): 자동차 이름 입력받는 기능 추가

hw0603 committed:
feat(validator): 자동차 이름 검증 메서드 구현

- 이름의 길이 초과 시 예외 발생
- 이름들 중 중복 발생 시 예외 발생
- 이름 중 영어가 아닌 문자 포함 시 예외 발생

hw0603 committed:
test(validatortest): 자동차 이름 검증 테스트 코드 작성

hw0603 committed:
feat(validator): 이동 횟수 검증 구현

hw0603 committed:
test(validator): 이동 횟수 검증 테스트 코드 작성

hw0603 committed:
feat(validator): 자동차 이름 검증 메소드 구현

hw0603 committed:
feat(inputview): 이동 횟수 입력 기능 구현

hw0603 committed:
feat(car): Car 클래스 생성자 추가

hw0603 committed:
feat(gamecontroller): 사용자 입력 및 자동차 목록 생성

hw0603 committed:
feat(randomNumberGenerator): 난수 생성 클래스 구현

hw0603 committed:
test(randomNumberGenerator): 생성된 난수 범위의 테스트 코드 작성

hw0603 committed:
fix(validatorTest): 접근제어자 수정에 따른 테스트 코드 수정

hw0603 committed:
feat(outputView): 자동차의 이름과 위치를 출력하는 메서드 작성

hw0603 committed:
feat(gameController): 한 라운드를 진행하는 메서드 구현

hw0603 committed:
feat(car): 자동차 위치를 반환하는 게터 추가

hw0603 committed:
feat(gameController): 입력된 횟수만큼 라운드 진행 메서드 추가

hw0603 committed:
fix(gameController): carList 재 선언 문제 수정

hw0603 committed:
fix(inputView): 안내 문구 출력 시 개행문자 추가

hw0603 committed:
refactor(outputView): 출력 메서드 오버로딩

hw0603 committed:
feat(gameController): 우승자 판별 및 출력 메서드 구현

hw0603 committed:
feat(application): 메인 메서드 구현

hw0603 committed:
feat(gameController): 우승자가 없는 경우 구현

- 이동한 자동차가 없을 경우 출력 메서드 추가

hw0603 committed:
docs(readme): 요구사항 명세 추가

hw0603 committed:
refactor(gameController): 최대 이동 위치과 우승자를 출력하는 메서드 추출

hw0603 committed:
refactor(car): 랜덤값을 인자로 받아 스스로 움직일지를 판단하도록 수정

- ValidatorTest 에 안쓰는 임포트문 제거

hw0603 committed:
refactor(car): 이름과 현재 위치를 반환하는toString() 구현

- 상수 포장

hw0603 committed:
refactor(gameController): 메서드 및 변수명 수정

hw0603 committed:
refactor(carGroup): Car 클래스 래핑

hw0603 committed:
style(gameController): 변수명 수정

hw0603 committed:
fix(gameController): finish 호출 위치 변경

- 이동 내역 출력 시 줄바꿈 수정

hw0603 committed:
test(carGroupTest): 우승자 판단 테스트 코드 추가

hw0603 committed:
test(carTest): 조건에 따른 이동 여부 테스트 코드 추가

hw0603 committed:
fix(carGroup): 움직인 자동차만 우승자가 될 수 있도록 수정

hw0603 committed:
refactor(car): 이동 기준값 상수 추출

hw0603 committed:
refactor(nameParser): 입력된 이름을 구분해 주는 클래스 추가

- 이름 목록을 String[] 에서 List<String> 으로 변경

hw0603 committed:
test(nameParserTest): 이름들을 구분하는 테스트 코드 추가

jminkkk committed:
feat(inputeView): 입력 횟수가 숫자가 아닌 경우 예외처리

jminkkk committed:
refactor(carGroup): List의 forEach 메서드 사용하여 변경

- throws IOException 제거
- 초기값 상수로 변경:

jminkkk committed:
refactor(gameController): 입력값이 유효한지를 나타내는 변수명 수정

jminkkk committed:
feat(moveHistory): 이동 내역을 저장하는 레코드 생성

jminkkk committed:
fix(carTest): car의 이름은 영어만 가능하기에 이름 수정

- 생성된 난수가 이동 기준을 만족하지 않는 경우, 난수룰 경계값으로 수정

jminkkk committed:
feat(randomNumberGenerator): 입력한 크기만큼의 난수를 생성하는 메서드 추가

- 입력한 크기만큼의 난수를 생성하는지 테스트하는 코드 추가

jminkkk committed:
refactor(validatorTest): Validator 클래스 삭제에 따른 검증 테스트 위치 변경

jminkkk committed:
refactor(inputView): 입력받은 이름을 파싱하여 리스트 형태로 반환하도록 수정

jminkkk committed:
rafactor(car): 검증 책임 부여 및 isWinner 메서드 추출

- equals 추가

jminkkk committed:
refactor(carGroup): 검증 책임 부여 및 경주 진행 시 MoveHistory를 반환하도록 변경

- race 시에 랜덤 숫자 리스트를 인자로 받도록 수정

jminkkk committed:
refactor(carGroupTest): race 시에 랜덤 값 목록을 인자로 받는다

jminkkk committed:
refactor(outputView): 이동 내역인 MoveHistory를 출력하는 메서드 추가

jminkkk committed:
refactor(gameController): 지역 변수 제거

jminkkk committed:
Merge branch 'jminkkk' into step2

ddaaac commented on src/main/java/racingcar/controller/GameController.java:
```diff
@@ -1,76 +1,54 @@
+        int moveCount = readMoveCount();
+        CarGroup carGroup = initCars(names);
 
-    public void init() {
-        readCarNames();
-        readMoveCount();
-        initCars(names);
```
리뷰를 하다보면 조금 마음에 걸리긴 하지만 지금 단계에서 굳이 이런거까지 리뷰해야할까 싶어서 넘기는 부분이 있습니다.
이 부분도 1단계 때 눈에 띄긴 했지만 넘겼던 부분인데 잘 수정해주셨네요.

ddaaac commented on src/main/java/racingcar/controller/GameController.java:
```diff
@@ -1,76 +1,54 @@
 
 public class GameController {
-    private final CarGroup carGroup = new CarGroup();
-    private List<String> names;
-    private int moveCount;
```
엄밀히 말하면 전역 변수는 아니긴 하죠. 용어로 표현하면 stateful한 상태라고 할 수 있을 거에요. 변경한 코드는 stateless라고 할 수 있구요.
이 부분도 잘 챙겨주셨습니다.

ddaaac commented on src/main/java/racingcar/model/MoveHistory.java:
```diff
@@ -0,0 +1,13 @@
+        for (String name : nameAndPosition.keySet()) {
+            sb.append(name).append(" : ").append("-".repeat(nameAndPosition.get(name))).append("\n");
+        }
+        return sb.toString();
```
> MoveHistory라는 Record를 통해 경주가 진행될 경우 각 자동차들의 이동 내역을 저장하자.
게임이 종료된 후에 모든 이동 내역을 결과와 함께 출력하도록 하는 것이 자연스러운 흐름같다.

History도 잘 두셨네요. 코드 흐름을 방해하지 않고 이전 결과를 모두 출력하게 잘 구현해주셨습니다.

다만, 출력에 관한 로직은 출력부에 두면 좋을 것 같아요. 단순히 데이터를 나르기만 해도 될 것 같네요.

ddaaac commented on src/main/java/racingcar/model/CarGroup.java:
```diff
@@ -1,25 +1,47 @@
+import java.util.Map;
 
 public class CarGroup {
-    private final List<Car> cars = new ArrayList<>();
+    private final List<Car> cars;
```
> 사실 "일급 컬렉션을 활용해야겠다" 하고 사용한 건 아니에요🥲
어쩌다보니 일급 컬렉션을 사용한 샘인데, 일급 컬렉션이 무엇인지에 학습해보면서 잘못 사용한 일급 컬렉션에 대해 고민해보았어요 :)

의도한 게 아니었군요 😮 
그렇다면 일급 컬렉션에 대해서는 다음 단계에서 자연스럽게 학습할 수 있으니 이번 단계에서는 이 정도가 있다 하고 넘어가셔도 좋을 것 같아요.

> 제가 이번에 이해한 일급 컬렉션은 불변하다는 장점이 있는 것인데요. 그렇다면 제가 기존에 구현한 addCar() 메서드가 있는 CarGroup은 일급 컬렉션으로 보기에 무리가 있는 것일까요?

사람마다 생각이 다를 수는 있겠지만 꼭 일급컬렉션을 쓸 필요는 없다고 생각해요. 가변으로 작성하는 편이 더 읽기 편하고 유지보수도 편하다고 생각이 들면 일급컬렉션을 사용하지 않아도 되는 거죠. 조금 답변 핀트가 엇나간 거 같긴 한데 일급 컬렉션에 집착할 필요가 없다는 얘기를 하고 싶었습니다.

jminkkk replied:
네네! 적절한 상황에 맞게 사용하는 것이 더 중요한 것 같네요! 
감사합니다 범블비 😀

ddaaac commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -3,13 +3,15 @@
     private static final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
 
-    public static String inputNames() {
+    public static List<String> inputNames() {
```
> List 으로 파싱까지의 책임을 View에서 맡도록 하자.
현재는 콘솔 입력이지만 요구사항이 변경될 경우, Controller에서 다른 형태의 입력값을 List를 처리하기 위해 로직을 변경해야 한다. 따라서 InputView에서 List로 반환하여 이러한 변경에 대처할 수 있도록 하자.

문제에 접근하는 방향성이 좋네요 👍

ddaaac commented on src/main/java/racingcar/utils/RandomNumberGenerator.java:
```diff
@@ -1,8 +1,17 @@
+import java.util.stream.IntStream;
 
 public class RandomNumberGenerator {
+    public static List<Integer> generate(int size) {
```
[정적 메소드](https://tecoble.techcourse.co.kr/post/2020-07-16-static-method/) 사용에 대해서 고민해보시면 좋을 것 같습니다.

jminkkk replied:
'RandomNumberGenerator'는 유틸리티 클래스로 사용하기 위해 정적 메서드를 사용했는데, CarGroup안에서 RandomNumberGenerator를 가지고 있어도 될 것 같다는 생각이 드네요...!

아직 어떤 경우에 정적 메서드를 적절하게 사용해야 하는지 대한 기준이 아직 명확하지 않은 것 같아요... 
더 고민해보겠습니다...!

ddaaac commented on src/main/java/racingcar/model/CarGroup.java:
```diff
@@ -1,25 +1,47 @@
+        }
+    }
+
+    public MoveHistory race(List<Integer> randomNumbers) {
```
> 내가 제어할 수 없는 RandomNumberGenerator를 CarGroup에서 가지게 되면 단위테스트를 하기에 힘들 수 있다. 최대한 모델에서 제어할 수 없는 부분을 몰아내자.

테스트를 용이하게 만드는 접근이 좋네요.
좀 더 고민해보자면 지금의 `MoveHistory race(List<Integer> randomNumbers)` 시그니처만 보고 randomNumbers의 의도를 빠르게 파악할 수 있을까를 고민해보면 좋을 것 같아요. 지금은 cars의 size와 randomNumbers의 크기가 같아야한다는 것도 객체 외부에서 알아야하구요. 
List 대신 인터페이스를 넣어주는 방식도 한 번쯤 고려해볼만 하겠네요.

jminkkk replied:
범블비의 리뷰처럼 외부에서 race를 사용할 때 randomNumbers의 의도를 파악하기 쉽지 않은 코드인 것 같아요...🥲
말씀해주신 인터페이스를 활용하는 방식에 대해 고민하고 적용해보겠습니다!

ddaaac commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -20,13 +24,36 @@ public int getPosition() {
         }
     }
+
+    public boolean isWinner(int maxPosition) {
+        return this.position > 0 && this.position == maxPosition;
+    }
```
- 여기 시그니처도 고민해보시면 좋겠네요. 파라미터로 넘어오는 값이 "max"일 필요가 있을까요?
- 추가된 메서드들에 대해 테스트도 챙겨주세요.

jminkkk replied:
위너의 조건이 변경될 경우 변수명을 수정해야 할 것 같네요...
변수명에 대해서 고민해볼게요! 
추가된 메서드 테스트도 추가하겠습니다!

ddaaac commented on src/test/java/racingcar/model/CarGroupTest.java:
```diff
@@ -1,50 +1,51 @@
-        carGroup.add(pedro);
-        carGroup.add(moly);
-        carGroup.add(neo);
+        int moveNumber = 4;
+        int notMoveNumber = 0;
```
요런 변수 표현 좋네요 👍

ddaaac reviewed (changes requested):
안녕하세요 몰리 👋 

2단계 구현도 잘 해주셨네요.

몇 가지 코멘트 남겼으니 확인해주세요.

jminkkk committed:
refactor(car): 우승자인지를 확인하는 메서드의 인자명을 winnerPosition로 변경

jminkkk committed:
test(car): 자동차의 현재 위치를 우승자의 위치와 비교하여 우승자를 판정하는 테스트 추가

jminkkk committed:
refactor(moveHistory, outputView): History 출력 로직의 위치를 이동

jminkkk committed:
Merge branch 'step2' of https://github.com/jminkkk/java-racingcar into step2

jminkkk committed:
rafactor(numberGenerator): NumberGenerator 추출 및 race()의 인자 NumberGenerator로 변경

- NumberGenerator 인터페이스 추가
- RandomNumberGenerator, FixedNumberGenerator 구현 및 사용 로직 변경

ddaaac commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -20,13 +24,36 @@ public int getPosition() {
 
     @Override
-    public String toString() {
-        return this.name + " : " + "-".repeat(this.position);
+    public boolean equals(Object o) {
```
equals / toString과 같은 관용적인 메서드들은 클래스 제일 아래에 위치시켜주세요.

ddaaac commented on src/test/java/racingcar/model/CarTest.java:
```diff
@@ -20,11 +21,34 @@ void moveTest() {
+    @Test
+    @DisplayName("자동차의 현재 위치를 우승자의 위치와 비교하여 우승자를 판정한다.")
+    void isWinnerTest() {
+        final Car car = new Car("moly");
+        final int movePosition = 4;
+        final int winnerPosition = 3;
+
+        car.move(movePosition);
+        car.move(movePosition);
+        car.move(movePosition);
```
이렇게 move를 통해 이동할 수도 있지만, `new Car("moly", 3)`처럼 아예 생성자에서 position을 받을 수도 있습니다. 
이러면 테스트를 위한 생성자가 생기는게 아닌가 하고 반문할수도 있을텐데요,
그런 관점보다는 Car 객체 관점에서 `new Car("moly", 3)`라는 생성자가 있어도 되는가/어색하지는 않은가 라는 관점에서 바라보면 좋을 것 같습니다.

ddaaac reviewed (approved):
안녕하세요 몰리 👋 

리뷰 반영 잘 해주셨습니다.

자동차 미션 고생 많으셨고, 다음 미션들도 화이팅하세요 💪
